internal/pkg/instrumentation/bpf/go.opentelemetry.io/otel/global/bpf/probe.bpf.c

```diff
@@ -0,0 +1,173 @@
+    void *span_name_ptr = get_argument(ctx, 4);
+    u64 span_name_len = (u64)get_argument(ctx, 5);
+    u64 span_name_size = MAX_SPAN_NAME_LEN < span_name_len ? MAX_SPAN_NAME_LEN : span_name_len;
+    bpf_probe_read(span_name.buf, span_name_size, span_name_ptr);
+
+    // Save the span name in map to be read once the Start function returns
+    void *context_ptr_val = get_Go_context(ctx, 3, 0, true);
+    void *key = get_consistent_key(ctx, context_ptr_val);
+    bpf_map_update_elem(&span_name_by_context, &key, &span_name, 0);
+    return 0;
+}
+
+// This instrumentation attaches uprobe to the following function:
+// func (t *tracer) Start(ctx context.Context, name string, opts ...trace.SpanStartOption) (context.Context, trace.Span)
+SEC("uprobe/Start")
+int uprobe_Start_Returns(struct pt_regs *ctx) {
+    // Get the span name passed to the Start function
+    void *context_ptr_val = get_Go_context(ctx, 3, 0, true);
+    void *key = get_consistent_key(ctx, context_ptr_val);
+    struct span_name_t *span_name = bpf_map_lookup_elem(&span_name_by_context, &key); 
+    if (span_name == NULL) {
+        return 0;
+    }
+    bpf_map_delete_elem(&span_name_by_context, &key);
+
+    u32 zero_span_key = 0;
+    struct otel_span_t *zero_span = bpf_map_lookup_elem(&otel_span_storage_map, &zero_span_key);
+    if (zero_span == NULL) {
+        return 0;
+    }
+
+    u32 otel_span_key = 1;
+    // Zero the span we are about to build, eBPF doesn't support memset of large structs (more than 1024 bytes)
+    bpf_map_update_elem(&otel_span_storage_map, &otel_span_key, zero_span, 0);
+    // Get a pointer to the zeroed span
+    struct otel_span_t *otel_span = bpf_map_lookup_elem(&otel_span_storage_map, &otel_span_key);
+    if (otel_span == NULL) {
+        return 0;
+    }
+
+    otel_span->start_time = bpf_ktime_get_ns();
+    copy_byte_arrays((unsigned char*)span_name->buf, (unsigned char*)otel_span->span_name.buf, MAX_SPAN_NAME_LEN);
+
+    // Get the ** returned ** context and Span (concrete type of the interfaces)
+    void *ret_context_ptr_val = get_argument(ctx, 2);
+    void *span_ptr_val = get_argument(ctx, 4);
+
+    struct span_context *span_ctx = get_parent_span_context(ret_context_ptr_val);
+    if (span_ctx != NULL) {
+        // Set the parent context
+        bpf_probe_read(&otel_span->psc, sizeof(otel_span->psc), span_ctx);
+        copy_byte_arrays(otel_span->psc.TraceID, otel_span->sc.TraceID, TRACE_ID_SIZE);
+        generate_random_bytes(otel_span->sc.SpanID, SPAN_ID_SIZE);
+    } else {
+        otel_span->sc = generate_span_context();
+    }
+
+    bpf_map_update_elem(&active_spans_by_span_ptr, &span_ptr_val, otel_span, 0);
+    start_tracking_span(ret_context_ptr_val, &otel_span->sc);
+    return 0;
+}
+
+// This instrumentation attaches uprobe to the following function:
+// func (*nonRecordingSpan) SetAttributes(...attribute.KeyValue)
+SEC("uprobe/SetAttributes")
+int uprobe_SetAttributes(struct pt_regs *ctx) {
```

Comment:
The following methods need to be included to provide a full implementation:

- `IsRecording`
  - This is critical given many users will use this method to gate other telemetry annotation operations.
- `SetStatus`
- `SetError`
- `RecordError`
- `AddEvent`
- `SetName`

The tracking of status and events will mean that the `probe.Event` needs to be updated to support these.

Reply:
> `IsRecording` should be probed to always return true?

That's a good question. As long as the auto-instrumentation `TracerProvider` sampler doesn't return `DROP` this should return `true`: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#sampler

Is there a way we can inject a sampling function to determine what the sampler will return?

Reply:
We can pass any relevant information from Go to eBPF that will be relevant to the sampling logic (but not code to execute).

Reply:
`SetStatus` tracking issue: https://github.com/open-telemetry/opentelemetry-go-instrumentation/issues/538

Reply:
`SetError` tracking issue: https://github.com/open-telemetry/opentelemetry-go-instrumentation/issues/539

Reply:
`RecordError` tracking issue: https://github.com/open-telemetry/opentelemetry-go-instrumentation/issues/540

Reply:
`AddEvent` tracing issue: https://github.com/open-telemetry/opentelemetry-go-instrumentation/issues/541

Reply:
`SetName` tracking issue: https://github.com/open-telemetry/opentelemetry-go-instrumentation/issues/542